src/mnode/mnode-pastel.cpp

```diff
@@ -1382,6 +1445,50 @@ std::vector<CArtTradeTicket> CArtTradeTicket::FindAllTicketByArtTnxID(const std:
     return masterNodeCtrl.masternodeTickets.FindTicketsByMVKey<CArtTradeTicket>(artTnxID);
 }
 
+std::vector<CArtTradeTicket> CArtTradeTicket::FindAllTicketByRegTnxID(const std::string& nftRegTnxId)
+{
+    return masterNodeCtrl.masternodeTickets.FindTicketsByMVKey<CArtTradeTicket>(nftRegTnxId);
+}
+
+std::map<std::string, std::string> CArtTradeTicket::GetPastelIdAndTxIdWithTopHeightPerCopy(const std::vector<CArtTradeTicket> & filteredTickets)
+{
+  //The list is already sorted by height (from beginning to end)
+
+  //This will hold all the owner / copies serial number where serial number is the key 
+  std::map<std::string, std::string> ownerPastelIDs_and_txids;
+
+  //Copy number and winning index (within the vector)
+  std::map<std::string, int> copyOwner_Idxs;
+  int winning_idx = 0;
+
+  for (const auto & element : filteredTickets) {
+
+    const std::string& serial = element.GetCopySerialNr();
+    if(copyOwner_Idxs.find(serial) != copyOwner_Idxs.end())
+    {
+      //We do have it in our copyOwner_Idxs
+      if(element.GetBlock() >= copyOwner_Idxs[serial])
```

Comment:
you save in 'copyOwner_Idxs' map value of 'filteredTickets' index . I can not understand why you compare CArtTradeTicket m_nBlock with saved index of element. If you have sorted filteredTickets from low to high blocks then probably you do not need this checking?

Reply:
No. Please check the logic you dont get the point ! There can be multiple copies. So the 'filtereTickets' just filtered for an art registration ticket. This means it can have hundreds of items in it. And each copy with the highest number of trade block (per each copy!!) has the 'owner'.

Reply:
Ok. let's consider filteredTickets has 3 items for the same copy number 3:
1 ArtTradeTicket with m_nBlock = 100
2 ArtTradeTicket with m_nBlock = 300
3 ArtTradeTicket with m_nBlock = 200

If look on your code ar first iteration copyOwner_Idxs will get insert [key=copyNumber 3, value = winning_idx 0]

at second iteration GetBlock with value 300 will be > copyOwner_Idxs[serial] with value 0, so copyOwner_Idxs will be updated to [key=copyNumber 3, value = winning_idx 1]

at third iteration GetBlock with value 200 will be > copyOwner_Idxs[serial] with value 1, so copyOwner_Idxs will be updated to [key=copyNumber 3, value = winning_idx 2]

Let's @a-ok123  @akobrin1 will look on this part of code.

Reply:
I suppose it should be like    
    std::map<std::string, std::pair<height, int>> copyOwner_Idxs;
    if(copyOwner_Idxs.find(serial) != copyOwner_Idxs.end())
    {
      //We do have it in our copyOwner_Idxs
      if(element.GetBlock() >= copyOwner_Idxs[serial].first)
      {
        copyOwner_Idxs[serial] = std::make_pair(element.GetBlock(), winning_idx);
      }
     else
    {
      copyOwner_Idxs.insert({ serial,  std::make_pair(element.GetBlock(), winning_idx) });
    }
    winning_idx++;

Reply:
Valid points! I'll modify it accordingly !